### ISSUE TITLE
New lint: Returning unit type when expecting Ord

### ISSUE THREAD

**Issue by Uriopass** [rust-lang/rust-clippy#5080]
I had a very weird bug today, I was trying to sort my list using a somewhat complicated key expression with `sort_by_key`, but no sorting was happening it was like getting a random order.  

Turns out I had an extra semicolon at the end of my expression, making my key return `()` instead which implements `Ord` (always `Equal`).

Example:

```Rust
vec.sort_by_key(|x| {
    some_function(some_expression(x));
                                  // ^---- extra semicolon here
});
```

**Comment by Areredify** [comment]
Should extend this to all places that expect `Fn(...) -> K, K: Ord` and that function is not a `|...| ()`?

**Comment by Uriopass** [comment]
Yes, that's pretty much what my titles says I think.

**Comment by Areredify** [comment]
It's probably not a good first issue, you have to look at the bounds of the generic function and check if the last statement of the closure is an explicit `()`, I would say it's too confusing

**Comment by flip1995** [comment]
Yeah, to make it generally apply to `Fn(...) -> K, K: Ord` types might be a little bit harder. But only linting `sort_by_key` is just calling `cx.tcx.expr_ty(last_expr_in_closure)`, and checking if it is `()`-type.

**Comment by Aaron1011** [comment]
I think this could be generalized to any function that implicitly returns `()`. If we have code like this:

```rust
fn my_fn() -> SomeType { 
    ...
}

fn foo<T>() -> T where T: SomeBound() {
    ...
    my_expr();
}
```

If both `SomeType` and `()` are valid return types for `foo` (that is, they both satisfy all of the bounds), we could suggest making the intended value more explicit (`return my_expr();` or `return ()`).

**Comment by rabisg0** [comment]
From what I've seen `return ()` doesn't remain in the hir so linting based on that might be difficult.
